rfc/20241023-rfc-tracking-issues.md

```diff
@@ -0,0 +1,209 @@
+
+    In particular, any pull requests that are intended as progress towards the completion of the features described in the RFC should be directly linked as checklist items so that there is a single place to find work that was already completed and work that's awaiting review, to help avoid accidental duplicate or conflicting work between people working asynchronously.
+
+An RFC tracking issue is tagged with the same "rfc" label we currently use for RFC pull requests, since the tracking issue represents the remaining work related to a specific RFC. [A label search for "rfc"](https://github.com/opentofu/opentofu/labels/rfc) would therefore return both the not-yet-approved RFCs represented as pull requests and the accepted-but-not-completed RFCs represented as tracking issues.
+
+An RFC tracking issue should also have exactly one of the following labels summarizing its current status:
+
+- "rfc-blocked": There are any blockers listed in the description that are not yet resolved, and so final implementation should not begin.
+- "rfc-ready": All blockers are resolved so implementation would could potentially begin, but implementation work is not yet assigned to anyone. Anyone in the community could potentially volunteer to work on an RFC in this stage, though they should discuss their intention first and wait for transition into the next phase before writing any code that's intended to be merged.
+- "rfc-implementing": One or more specific individuals have signed up to implement what's described in the RFC and are actively working on it. At this stage the tracking issue should also be assigned to each of the assigned individuals.
+
+While the issue is in the "ready" or "implemented" phases it's possible that the folks working on it will learn of new blockers that were not originally anticipated, in which case they can make a judgement call about whether the blockers are significant enough to return to the "rfc-blocked" state or if they can be resolved promptly.
+
+The tracking issue conventionally "belongs to" some specific individual who is acting as a project owner for the implementation. In the short term that is likely to be a core team member, but in the long run could potentially be a community member who is willing and able to take on that role. That individual is ultimately responsible for keeping the main issue description and labels updated as the status changes, which implies that the individual must either be the author of the issue or must have "triage" access to the repository under [GitHub's access control scheme](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization). In practice the project owner may of course delegate some or all of these responsibilities to others; the requirement is only that there be a single person accountable for knowing the overall status of the work and able to communicate that with other stakeholders.
```

Comment:
Totally agreed about linking other issues. I must admit I hadn't really considered the idea of using _issues_ as "subtasks" -- my earlier writing talked about explicitly listing out all of the _pull requests_ because I imagined that as being the primary way to represent specific subtasks.

But I can imagine situations where it could be helpful to create an issue representing a subtask that is itself complex enough to warrant multiple pull requests, in which case linking those would be important too.

Are you okay with letting this RFC remain a little vague on this for now, and then learning from what happens in practice on real projects we run in the near future? I was intentionally hesitating on being _too_ prescriptive about exactly how the individual work items ought to be represented and trying to focus on the general idea that _the tracking issue must link to everything else_ regardless of what the "everything else" might be for a particular project, since I expect projects of different sizes and shapes would have different needs.

Reply:
I wonder if we should expand the owner / "project manager" responsibility also to be the technical lead and product lead of the RFC.
As RFCs change over time, and while working on them, people might hit blocks that require changing the original RFC. These blocks might be actual technical blocks that we missed in the RFC or just blocks that will be too costly, and we can decide to change the technical implementation or the user-facing behavior to avoid facing them.
So, as the technical / product lead, the owner will be responsible for keeping the RFC up to date, and he'll challenge any possible alterations (technical or user-facing) during the implementation.
He doesn't have to update the RFC on his own, but he should discuss the intended changes, provide input, and oversee that the agreed-upon changes are merged back into the RFC.
It's essential to have this persona, as he took the time to investigate and understand all the trade-offs of his suggested RFC, while the people working on the RFC might not be that versed in the details and all the different considerations.

Reply:
I saw you addressed most of the operative part of what I mentioned above in the "RFC Amendments". But it'll still be interesting to hear your opinion about the RFC owner being the technical lead / product lead of the RFC implementation.

Reply:
I had intended this RFC to focus primarily on the communication about the project rather than on other project management details, but I do agree with you that the person who is responsible for keeping the tracking issue updated is probably also the main person responsible for ensuring the RFC stays updated and that the final implementation actually does match the RFC.

I'd hesitate to grow the scope of this RFC to talk about project management _in general_; my intention had only been to prescribe a way to communicate about a project in a way that's compatible with asynchronous and decentralized work and external observers. But I do think that we could separately agree within the core team that by default we have a single lead that's responsible for all of these roles, unless a particular project decides to split that work up differently for some special reason.

Reply:
It makes sense to me; I agree that we shouldn't include it in the scope of this RFC.